cloud_Azure/terraform/module/examples/multiple_accounts_multiple_resource_group/profiles_populator.py

```diff
@@ -0,0 +1,479 @@
+    except KeyError:
+        log.exception("Error parsing returned Service Principal")
+        return None
+
+
+def add_read_write_all_permissions(principal_id: str) -> bool:
+    COMMAND = f"ad app permission add --id {principal_id} --api-permissions {AZURE_READ_WRITE_ALL_PERMISSION} --api {AZURE_GRAPH_API}"  # returns a dict
+    ATTEMPT_COUNT = 5
+    output_dict = az_cli(COMMAND, ATTEMPT_COUNT)
+    if not isinstance(output_dict, dict):
+        log.error("Failed to give read-write-all permissions to Service Principal")
+        return False
+    return True
+
+
+def grant_permissions(principal_id: str) -> bool:
+    COMMAND_GRANT = f"ad app permission grant --id {principal_id} --api {AZURE_GRAPH_API}"  # returns a dict
+    ATTEMPT_COUNT = 5
+    output_dict = az_cli(COMMAND_GRANT, ATTEMPT_COUNT)
+    if not isinstance(output_dict, dict):
+        log.error("Failed to grant permissions to Service Principal")
+        return False
+    return True
+
+
+def consent_permissions(principal_id: str) -> bool:
+    COMMAND = f"ad app permission admin-consent --id {principal_id}"  # returns a dict
+    ATTEMPT_COUNT = 5
+    output_dict = az_cli(COMMAND, ATTEMPT_COUNT)
+    if not isinstance(output_dict, dict):
+        log.error("Failed to consent permissions to Service Principal")
+        return False
+    return True
+
+
+def save_profiles(file_path: str, profiles: List[AzureProfile]) -> bool:
+    config = configparser.ConfigParser()
+    for profile in profiles:
+        section = profile.name
+        config.add_section(section)
+        config[section]["subscription_id"] = profile.subscription_id
+        config[section]["tenant_id"] = profile.tenant_id
+        config[section]["principal_id"] = profile.principal_id
+        config[section]["principal_secret"] = profile.principal_secret
+        config[section]["location"] = profile.location
+        config[section]["resource_group_names"] = ",".join(profile.resource_group_names)
+        config[section]["storage_account_names"] = ""
+
+    try:
+        with open(file_path, "w", encoding="utf-8") as configfile:
+            config.write(configfile)
+        log.info("Saved profiles to '%s'", file_path)
+        return True
+    except OSError:
+        log.exception("Failed to save profiles to '%s'", file_path)
+        return False
+
+
+def profile_exists(profile_name: str, profiles: List[AzureProfile]) -> bool:
+    return find_profile(profile_name, profiles) is not None
+
+
+def find_profile(profile_name: str, profiles: List[AzureProfile]) -> Optional[int]:
+    for i, profile in enumerate(profiles):
+        if profile.name == profile_name:
+            return i
+    return None
+
+
+def find_secret(principal_id: str, profiles: List[AzureProfile]) -> Optional[str]:
+    for profile in profiles:
+        if profile.principal_id == principal_id and likely_valid_service_principal_secret(profile.principal_secret):
```

Comment:
Agreed (assuming that "33 signs" means "33 characters").

Reply:
Added "Invalid secret" warning

Reply:
> > Note: If we allow the user to fill profiles manually, it might make sense to validate the inputted data, eg check if inputted "location" exists in azure, check if inputted "resource groups" exist in specified location. Validation would be next step after filling missing data. To be considered.
> 
> Agreed. With invalid data in `profiles.ini` `azure_onboarder.py` will simply fail to create resources for the offending profile, correct? How obvious are error messages that it produces?

How obvious they are, I think it depends on the operator's knowledge regarding Terraform AWS provider and error logs